---
title: "At_vs_Zm_Seurat_CCA"
output: html_document
---


```{r}

library("rlang")
library("Seurat");
library("sctransform");
library("dplyr");
library("RColorBrewer");
library("ggthemes");
library("ggplot2");
library("cowplot");
library("data.table");

```

##### section-1 convert ortholog of maize gene IDs to arabidopsis for 1 to 1 ortholog only #########

```{r}
## follow link below to perform the ortholog convertion 
## https://github.com/mamtagiri/SingleCellSequencinginR/blob/master/OrthologsConvertion.R

##########################################
# Ortholog gene name conversion in seurat
##########################################
library(Seurat)
library(dplyr)

# Add the path to the appropriate rds file
# read the maize rds file after seurat clustering and set asssay to RNA

setwd("C:/Users/Xiaosa Xu/Desktop/SC Stem Cell Manuscript/Seurat_sc_datasets/1_B73_ear_tip_xu16192932_Dec2022_integration")

B73.tip <- readRDS(file = "B73_ear_tip.int.subcluster6.rename.rds")

DefaultAssay(B73.tip)<-"RNA"


# read the Arabidopsis rds file after seurat clustering and set asssay to RNA

setwd("C:/Users/Xiaosa Xu/Desktop/SC Stem Cell Manuscript/Seurat_sc_datasets/2_ap1cal_xu41_42_Dec2022_integration")

ap1cal.int <- readRDS(file = "ap1cal.int_UMI-1000-100000_Gene-1000-12000.rds")

DefaultAssay(ap1cal.int)<-"RNA"


#read the orthology file, add path to the file
 
#use Michael truepair 80_tolerance
setwd("C:/Users/Xiaosa Xu/Desktop/SC Stem Cell Manuscript/Seurat_sc_datasets/3_B73_ear_tip_vs_ap1cal-rerun-Jan2024")

trupair <- read.table("true_pair_dataframe_at_80_tolerance_jack.txt", sep = "\t", header = TRUE)

#subset data to the genes with orthologs
# create a list of all the gene names in the above step (ortholog pairs)


#use Michael truepair 80_tolerance
maizegenes<- as.vector(unlist(trupair$Final.V3.IDs.in.Maize.Seurat.Feature))
head(maizegenes)

arabigenes<- as.vector(unlist(trupair$Arabidopsis.Gene))
head(arabigenes)


# subsetting the rds file to only genes that have orthologous pairs, you can skip this step to keep all genes

maize_ortho<-subset(B73.tip, features = maizegenes)
maize_ortho


arabi_ortho<-subset(ap1cal.int, features = arabigenes)
arabi_ortho


###Note: it seems maize_ortho and arabi_ortho has different number of features after subseting using trupair. This is because not all genes from truepairs are expressed. 

#convert gene names for maize data
# iterate over each maize gene name, find its position and corresponding ortholog name from the "true_pair" file. Search the maize name in the rds file and change to the ortholog just read.  

 
 #use Michael truepair 80_tolerance

for (i in maizegenes){
  pn <-paste0("^",i,"$")  # using regex for exact match
  p=which(grepl(pn,trupair$Final.V3.IDs.in.Maize.Seurat.Feature)) # find position of the match
  p1<-p[1]  # keeping the first position from p
  f<-trupair$Arabidopsis.Gene[p1] # get the arabidopsis gene name from the same position
  pn2<-i
  pn3<-paste0("^",pn2,"$") # regex exact match for the maize name
  k=grep(pn3,maize_ortho@assays$RNA@counts@Dimnames[[1]]) # grep the position in the rds maize object for the maize gene name
  k1<-k[1] # keep only the first position from above
  maize_ortho@assays$RNA@counts@Dimnames[[1]][k1]<-as.character(f)} # change the exact position with its correspondong arabidopsis gene name read above

 
#save the rds with gene names changed--use Michael truepair
saveRDS(maize_ortho,file="maizeOrthologs_80_tolerance_rerun.rds")

saveRDS(arabi_ortho,file="arabiOrthologs_80_tolerance_rerun.rds")


## export the raw count for maize_ortho

count<- as.matrix(GetAssayData(object = maize_ortho, slot = "counts"))

write.table(count, file = "maize_ortho_Avg_count.txt")


## export the raw count for arabi_ortho

count<- as.matrix(GetAssayData(object = arabi_ortho, slot = "counts"))

write.table(count, file = "arabi_ortho_Avg_count.txt")


### check gene IDs
rownames(maize_ortho@assays$RNA)
rownames(arabi_ortho@assays$RNA)
```


```{r}
## create new object


arabi_genes <- CreateSeuratObject(counts = arabi_ortho@assays$RNA@counts, assay= "RNA") 

arabi_genes 



maize_genes <- CreateSeuratObject(counts = maize_ortho@assays$RNA@counts, assay= "RNA")

maize_genes 


```


##### section-3 The standard Seurat integration pipeline, which also uses CCA#########


```{r}
DefaultAssay(arabi_genes)<-"RNA"

DefaultAssay(maize_genes)<-"RNA"


arabi_genes@meta.data$source <- "arabi"
maize_genes@meta.data$source <- "maize"


At.Zm.list = list()

At.Zm.list[[1]] <- subset(arabi_genes)

At.Zm.list[[2]] <- subset(maize_genes)


```


```{r}

memory.limit(size=50000000)

## logNormalize


At.Zm.list[[1]]<- NormalizeData(At.Zm.list[[1]], normalization.method = "LogNormalize", scale.factor = 10000)


At.Zm.list[[2]]<- NormalizeData(At.Zm.list[[2]], normalization.method = "LogNormalize", scale.factor = 10000)


```


```{r}
##Find variables

At.Zm.list[[1]]<- FindVariableFeatures(At.Zm.list[[1]], selection.method = "vst", nfeatures = 2000)

At.Zm.list[[2]]<- FindVariableFeatures(At.Zm.list[[2]], selection.method = "vst", nfeatures = 2000)

##Select Integration Features

features <- SelectIntegrationFeatures(object.list = At.Zm.list)

#Find anchors

anchors <- FindIntegrationAnchors(object.list = At.Zm.list, anchor.features = features)

#Integrate data

At.Zm.int <- IntegrateData(anchorset = anchors, dims = 1:30)

```



```{r}
DefaultAssay(At.Zm.int)<-"integrated"


# Run the standard workflow for visualization and clustering
At.Zm.int <- ScaleData(At.Zm.int)
At.Zm.int <- RunPCA(At.Zm.int)


```


```{r}
## Determine the ‘dimensionality’ of the dataset

ElbowPlot(At.Zm.int, ndims = 40)

## Elbow plot: quantitative approach
## https://hbctraining.github.io/scRNA-seq/lessons/elbow_plot_metric.html

# Determine percent of variation associated with each PC
pct <- At.Zm.int[["pca"]]@stdev / sum(At.Zm.int[["pca"]]@stdev) * 100

# Calculate cumulative percents for each PC
cumu <- cumsum(pct)

# Determine which PC exhibits cumulative percent greater than 90% and % variation associated with the PC as less than 5
co1 <- which(cumu > 90 & pct < 5)[1]

co1

# Determine the difference between variation of PC and subsequent PC
co2 <- sort(which((pct[1:length(pct) - 1] - pct[2:length(pct)]) > 0.1), decreasing = T)[1] + 1

# last point where change of % of variation is more than 0.1%.
co2


# Minimum of the two calculation
pcs <- min(co1, co2)

pcs

```



```{r}
DefaultAssay(At.Zm.int)<-"integrated"

At.Zm.int <- RunTSNE(At.Zm.int, reduction = "pca", dims = 1:13)

At.Zm.int <- FindNeighbors(At.Zm.int, reduction = "pca", dims = 1:13)

At.Zm.int <- FindClusters(At.Zm.int, resolution = 1.6)

```

```{r}
# Visualization
p1 <- DimPlot(At.Zm.int, reduction = "tsne", group.by = "source", pt.size = 0.5)
p2 <- DimPlot(At.Zm.int, reduction = "tsne", label = TRUE, repel = TRUE, pt.size = 0.5, label.size = 6)
p1 + p2


DimPlot(At.Zm.int, reduction = "tsne", split.by = "source", pt.size = 1)


```


```{r}

# Generate a heatmap for both conserved markers and species-specific markers in a specific cluster

# Load necessary libraries
library(Seurat)
library(pheatmap)
library(metap)
library(multtest)
library(BiocManager)

# Set the default assay to RNA for species-specific markers
DefaultAssay(At.Zm.int) <- 'RNA'

# Set the cluster identities using 'integrated_snn_res.1.6'
Idents(At.Zm.int) <- "integrated_snn_res.1.6"

# --- Step 1: Identify conserved marker genes in cluster 1 ---
# Switch to integrated assay to find conserved markers
DefaultAssay(At.Zm.int) <- 'integrated'

# Find conserved markers for cluster 1 without min.pct
conserved.markers.cluster1 <- FindConservedMarkers(At.Zm.int, 
                                                   ident.1 = 1,  # for cluster 1
                                                   grouping.var = "source", 
                                                   logfc.threshold = 0.25)  # log fold change cutoff

# Filter conserved markers by p_val_adj < 0.01
conserved.markers.cluster1.filtered <- conserved.markers.cluster1[
  conserved.markers.cluster1$max_pval < 0.01, ]

# Save conserved markers to CSV
write.csv(conserved.markers.cluster1.filtered, file = "ConservedMarkers_Cluster1.csv")

# --- Step 2: Identify species-specific markers for cluster 1 ---

# Set back to RNA assay for species-specific marker detection
DefaultAssay(At.Zm.int) <- 'RNA'

# Find Arabidopsis-specific markers in cluster 1
arabidopsis.markers.cluster1 <- FindMarkers(At.Zm.int, 
                                            ident.1 = "arabi",  # Arabidopsis
                                            ident.2 = "maize",  # Maize
                                            subset.ident = "1",  # for cluster 1
                                            group.by = "source", 
                                            logfc.threshold = 1)

# Filter Arabidopsis-specific markers by p_val_adj < 0.01
arabidopsis_enriched_cluster1 <- arabidopsis.markers.cluster1[
  arabidopsis.markers.cluster1$avg_log2FC > 0 & arabidopsis.markers.cluster1$p_val_adj < 0.01, ]

# Save Arabidopsis-specific markers to CSV
write.csv(arabidopsis_enriched_cluster1, file = "Arabidopsis_Enriched_Markers_Cluster1.csv")


# Find Maize-specific markers in cluster 1
maize.markers.cluster1 <- FindMarkers(At.Zm.int, 
                                      ident.1 = "maize",  # Maize
                                      ident.2 = "arabi",  # Arabidopsis
                                      subset.ident = "1",  # for cluster 1
                                      group.by = "source", 
                                      logfc.threshold = 0.25)

# Filter Maize-specific markers by p_val_adj < 0.01
maize_enriched_cluster1 <- maize.markers.cluster1[
  maize.markers.cluster1$avg_log2FC > 0 & maize.markers.cluster1$p_val_adj < 0.01, ]

# Save Maize-specific markers to CSV
write.csv(maize_enriched_cluster1, file = "Maize_Enriched_Markers_Cluster1.csv")


# --- Step 3: Generate Heatmap Similar to Figure 5d ---

# Combine marker lists for heatmap (you may want to focus on top markers)
combined_markers <- unique(c(rownames(conserved.markers.cluster1.filtered),
                             rownames(arabidopsis_enriched_cluster1),
                             rownames(maize_enriched_cluster1)))

# Extract the expression data for these markers from the RNA assay
DefaultAssay(At.Zm.int) <- 'RNA'
expression_data <- FetchData(At.Zm.int, vars = combined_markers)

# Normalize the data for better visualization (optional)
scaled_expression_data <- scale(t(expression_data))

# Create the heatmap
pheatmap(scaled_expression_data, 
         cluster_rows = TRUE, 
         cluster_cols = TRUE, 
         show_rownames = TRUE, 
         show_colnames = FALSE, 
         fontsize_row = 8, 
         scale = "row")

# Save the heatmap to a file (optional)
# pdf("Cluster1_Conserved_and_Specific_Markers_Heatmap.pdf")
# pheatmap(scaled_expression_data)
# dev.off()


```
